### PR TITLE
VideoCommon: Fix SSBO layout and remove associated "bug"

### PR DESCRIPTION
**Pull request by tellowkrinkle**
Bug was ours, not AMD's

If you don't specify a layout, the driver's free to do whatever it wants and we're supposed to ask it what it did.  AMD probably went for std140 layout (where int[4] would have each int padded to 16 bytes)

I don't see any reason we'd want an SSBO without std430, so I stuck it into the macro, but we could put it outside like we do for UBOs if you'd prefer that

### CODE REVIEW

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Comment by Pokechu22** [comment]
Unfortunately my old AMD laptop is dead, so I don't have the ability to test this (unlike I did with #9764). This sounds plausible though.

**Comment by tellowkrinkle** [comment]
I can confirm that on a Radeon Pro 5600M using AMD's 20.45.40.15 driver (which is the newest offered to Intel Macs running windows), removing the bug entry breaks bbox and adding std430 fixes it

**Comment by iwubcode** [comment]
I can confirm this works on my AMD card on Windows.

I would approve this but the `std430` layout seems to require OpenGL 4.3.  Our minimum supported version is 3.0 AFAICT.  So there are quite a few places this would cause a problem.


As an aside, I'd love to update our minimum requirement to something newer..

**Comment by mbc07** [comment]
Wouldn't be possible to use the `std430` layout by default but somehow fall back to the `int4` approach if OpenGL 4.3 isn't available?

**Comment by iwubcode** [comment]
> Wouldn't be possible to use the std430 layout by default but somehow fall back to the int4 approach if OpenGL 4.3 isn't available?

Yeah something like that would be possible, it's not being done though.

The current way it would be done is a flag that says whether the feature is available.  Then that flag would be set to true if OpenGL 4.3 or higher exists.  If it's true you'd use the `std430`, if it wasn't then you'd use the other approach.

We already have many examples of this.  It's just very klunky to have so many checks around the code base.  That's why it would be great if we could update our minimum version..

**Comment by tellowkrinkle** [comment]
[The document on the ssbo extension](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_shader_storage_buffer_object.txt) specifically mentions std430, so I assume anything that supports ssbos also supports std430 layout?

**Comment by iwubcode** [comment]
Yes that's reasonable but you need to check for `bSupportsFragmentStoresAndAtomics` before enabling it.

File: Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp [hunk 1]
```diff
@@ -790,15 +790,15 @@ void ProgramShaderCache::CreateHeader()
...
+          "#define SSBO_BINDING(x) layout(std430, binding = x)\n"
           "#define IMAGE_BINDING(format, x) layout(format, binding = x)\n" :
           "#define ATTRIBUTE_LOCATION(x)\n"
           "#define FRAGMENT_OUTPUT_LOCATION(x)\n"
           "#define FRAGMENT_OUTPUT_LOCATION_INDEXED(x, y)\n"
           "#define UBO_BINDING(packing, x) layout(packing)\n"
           "#define SAMPLER_BINDING(x)\n"
           "#define TEXEL_BUFFER_BINDING(x)\n"
-          "#define SSBO_BINDING(x)\n"
+          "#define SSBO_BINDING(x) layout(std430)\n"
```
Comment (iwubcode) [inline Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp:801]: This is the else statement, where `layout` isn't supported by the language, so this needs to be removed

File: Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp [hunk 2]
```diff
@@ -790,15 +790,15 @@ void ProgramShaderCache::CreateHeader()
           "#define UBO_BINDING(packing, x) layout(packing, binding = x)\n"
           "#define SAMPLER_BINDING(x) layout(binding = x)\n"
           "#define TEXEL_BUFFER_BINDING(x) layout(binding = x)\n"
-          "#define SSBO_BINDING(x) layout(binding = x)\n"
+          "#define SSBO_BINDING(x) layout(std430, binding = x)\n"
```
Comment (iwubcode) [inline Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp:793]: This part needs to be checked against `bSupportsFragmentStoresAndAtomics ` probably similar to below

**Comment by iwubcode** [comment]
@tellowkrinkle - yes, however, we only enable that extension if `bSupportsFragmentStoresAndAtomics` is defined.  That also sets the flag on whether we support BBOX which I believe is where we use SSBOs.

So I imagine we only want to set `std430` if that flag is set.  Not sure what would happen if we don't on those systems, maybe a compiler error?

File: Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp (hunk 1 above)
Comment (tellowkrinkle) [reply Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp:801]: That's just a define, which won't be expanded unless we attempt to use an SSBO

Or are you trying to say that bbox can be enabled even without SSBO support?

**Comment by tellowkrinkle** [comment]
`bSupportsFragmentStoresAndAtomics` *is* the "SSBO supported" flag https://github.com/dolphin-emu/dolphin/blob/23ed611077d65b43d9ef8f0971e9e5c57699fd91/Source/Core/VideoBackends/OGL/OGLRender.cpp#L424-L425

Therefore
1. If a driver doesn't support SSBOs, then `bSupportsFragmentStoresAndAtomics` will be `false`
2. If `bSupportsFragmentStoresAndAtomics` is false, bbox will be disabled
3. If bbox is disabled, we will not put SSBOs into any glsl shaders
4. If we don't put SSBOs into any glsl shaders, `SSBO_BINDING(n)` will also not appear in any shaders (where `n` is a number)
5. If `SSBO_BINDING(n)` does not appear in any shaders, the `#define SSBO_BINDING(x) layout(std430)` will never be expanded anywhere
6. If the define isn't expanded, the glsl compiler will never be asked to compile a shader that uses `layout(std430)`

**Comment by iwubcode** [comment]
I had assumed that just having the `std430` in the shader header would be enough to be problematic but reading the spec, the `#define` operates similarly to a `#define` in C, so like you say it won't be expanded.

**Review by iwubcode — APPROVED** [review]
Sorry @tellowkrinkle , I should have followed my train of thought all the way through but I was running out the door.  Anyway, this LGTM and it's been tested.

**Comment by tellowkrinkle** [comment]
> I had assumed that just having the `std430` in the shader header would be enough to be problematic

I mean, for all the compiler knows, I'm going to expand it into a function after `uint std430 = 4;`
Unless they've reserved all identifiers starting with `std` or something, that should be perfectly acceptable

**Pull request merged**